### PR TITLE
ngUpgrade (AOT) - Add support for `require` and more lifecycle hooks

### PR DESCRIPTION
**Pull request by gkalpak**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

```
[x] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the new behavior?**
- Add support for `require` (as string, array, object).
- Add support for more lifecycle hooks: `$onDestroy`, `$postLink`
- More tests: deep nesting ng1/ng2, lifecycle hooks, `require`.
- Some refactorings (most notably 35aee1f) and fixes/changes.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

**Other information**:
Currently, `ngUpgrade` deviates from "native" ng1 is the following ways:
- Does not support "remote" templates (i.e. downloading templates at runtime).
- Does not support compile/pre-/post-link in directives.
- Calls `$onChanges` for `=` bindings too (ng1 only calls it for `@`/`<`).

### CODE REVIEW

**Comment by gkalpak** [comment]
This now ready for review. The travis failures are not related to this PR afaict.
/cc @petebacondarwin, @mhevery, @alexeagle

File: modules/@angular/upgrade/src/angular_js.ts [hunk 1]
```diff
@@ -72,7 +73,7 @@ export interface IDirective {
   terminal?: boolean;
   transclude?: boolean|'element'|{[key: string]: string};
 }
-export type DirectiveRequireProperty = Ng1Token[] | Ng1Token | {[key: string]: Ng1Token};
+export type DirectiveRequireProperty = string | string[] | {[key: string]: string};
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/angular_js.ts]: why this change?

File: modules/@angular/upgrade/src/aot/upgrade_component.ts [hunk 2]
```diff
@@ -165,6 +170,17 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck {
     });
   }
 
+  ngOnDestroy() {
+    this.callLifecycleHook('$onDestroy', this.controllerInstance);
+    // QUESTION: Who should be in charge of destroying `$componentScope`?
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/aot/upgrade_component.ts]: It should get destroyed automatically when the new scope that was created for the upgraded component is destroyed, which I assume is what causes `ngOnDestroy` to be called.

File: modules/@angular/upgrade/src/aot/upgrade_component.ts [hunk 3]
```diff
@@ -182,6 +198,28 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck {
...
+        }
+      });
+    }
+
+    return require;
+  }
+
+  private isRequireObject<T>(require: T|T[]|{[key: string]: T}): require is {[key: string]: T} {
+    return require && !Array.isArray(require) && typeof require === 'object';
+  }
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/aot/upgrade_component.ts]: Ooh some serious TypeScript foo going on here.

**Review by petebacondarwin — COMMENTED** [review]
Made a few initial comments

**Review by petebacondarwin — DISMISSED** [review]

**Comment by petebacondarwin** [comment]
first commit 89a534d8347aafeae2bb4c654aeb5029b1c8de11 LGTM

**Comment by petebacondarwin** [comment]
second commit 89a534d8347aafeae2bb4c654aeb5029b1c8de11 LGTM

File: modules/@angular/upgrade/src/aot/upgrade_component.ts [hunk 4]
```diff
@@ -189,6 +198,28 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck {
     return directive;
   }
 
+  private getDirectiveRequire(directive: angular.IDirective): angular.DirectiveRequireProperty {
+    const require = directive.require || (directive.controller && directive.name);
+
+    if (this.isRequireObject<string>(require)) {
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/aot/upgrade_component.ts]: The string cast is not needed here - it is implicit in the function call.

File: modules/@angular/upgrade/src/aot/upgrade_component.ts (hunk 3 above)
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/aot/upgrade_component.ts]: Perhaps this should be a more generic helper?
E.g. 
````
function isPlainObject<T>(x: any): x is {[key: string]: T} {
  return x && !Array.isArray(x) && typeof x === 'object';
}
```

This could be moved to `facade/lang` if accepted. And it also means that both casts, even the `<IControllerInstance>` one, can be removed.

File: modules/@angular/upgrade/src/aot/upgrade_component.ts [hunk 5]
```diff
@@ -267,9 +298,48 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck {
     return controller;
   }
 
-  private resolveRequired(
-      $element: angular.IAugmentedJQuery, require: angular.DirectiveRequireProperty) {
-    // TODO
+  private resolveRequire(
+      directiveName: string, $element: angular.IAugmentedJQuery,
+      require: angular.DirectiveRequireProperty): IControllerInstance|IControllerInstance[]|
+      {[key: string]: IControllerInstance} {
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/aot/upgrade_component.ts]: Perhaps we could come up with a generic interface type that encapsulates `T|T[]|{[key:string]: T}`?

File: modules/@angular/upgrade/src/aot/upgrade_component.ts [hunk 6]
```diff
@@ -267,9 +298,48 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck {
...
+    if (!require) {
+      return null;
+    } else if (Array.isArray(require)) {
+      return require.map(req => this.resolveRequire(directiveName, $element, req));
+    } else if (typeof require === 'object') {
+      const value: {[key: string]: IControllerInstance} = {};
+
+      Object.keys(require).forEach(
+          key => { value[key] = this.resolveRequire(directiveName, $element, require[key]); });
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/aot/upgrade_component.ts]: I feel like we ought to have a helper that maps properties from one object to another...

E.g. https://lodash.com/docs/4.16.4#mapValues

**Comment by petebacondarwin** [comment]
third commit 4f21230457ee8a0bfe8270f7750b4891a877c0a3 LGTM except for a few potential refactorings

**Comment by petebacondarwin** [comment]
fourth commit 5a472ad5fa6a157b99733826a70ae47228a888ee LGTM except for type change from `Ng1Token`

File: modules/@angular/upgrade/test/aot/integration/upgrade_component_spec.ts [hunk 7]
```diff
@@ -537,11 +543,11 @@ export function main() {
            const element = html(`<ng2></ng2>`);
 
            platformBrowserDynamic().bootstrapModule(Ng2Module).then(ref => {
-             var adapter = ref.injector.get(UpgradeModule) as UpgradeModule;
+             const adapter = ref.injector.get(UpgradeModule) as UpgradeModule;
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/test/aot/integration/upgrade_component_spec.ts]: +1 :-)

Comment (petebacondarwin) [inline modules/@angular/upgrade/test/aot/integration/upgrade_component_spec.ts]: +1 :-)

**Comment by petebacondarwin** [comment]
fifth commit 21c9a70b35465603803dd240c2c88bc6dab051da LGTM

File: modules/@angular/upgrade/test/aot/integration/upgrade_component_spec.ts [hunk 8]
```diff
@@ -1315,100 +1312,432 @@ export function main() {
...
-            //   adapter.bootstrap(element, ['ng1']).ready((ref) => {
-            //     expect(multiTrim(document.body.textContent)).toEqual('PARENT:WORKS');
-            //     ref.dispose();
-            //   });
-            // }));
-        });
+    describe('require', () => {
+      // NOT YET SUPPORTED
+      xdescribe('in pre-/post-link', () => {
+        it('should resolve to its own controller if falsy', async(() => {
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/test/aot/integration/upgrade_component_spec.ts]: Is this what A1 does? I always thought you still had to manually require the current directive's controller if you wanted it.

**Comment by petebacondarwin** [comment]
sixth commit 6de2977bdda1c05994f59cdfda10ffd30b014b8f LGTM

File: modules/@angular/upgrade/test/aot/integration/upgrade_component_spec.ts [hunk 9]
```diff
@@ -1804,11 +1804,8 @@ export function main() {
            // Bootstrap
            const element = html(`<ng2></ng2>`);
 
-           platformBrowserDynamic().bootstrapModule(Ng2Module).then(ref => {
-             var adapter = ref.injector.get(UpgradeModule) as UpgradeModule;
-             adapter.bootstrap(element, [ng1Module.name]);
-
-             expect(multiTrim(document.body.textContent)).toBe('Called: yes | Called: yes');
+           bootstrap(platformBrowserDynamic(), Ng2Module, element, ng1Module).then(() => {
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/test/aot/integration/upgrade_component_spec.ts]: +1

**Comment by petebacondarwin** [comment]
seventh commit 3f1a2446f3a8ce84fe459f3dd6977c9ca7919b84 LGTM

File: modules/@angular/upgrade/test/aot/integration/upgrade_component_spec.ts [hunk 10]
```diff
@@ -1740,8 +1740,156 @@ export function main() {
...
+           // Define `ng1Directive`
+           const ng1DirectiveA: angular.IDirective = {
+             template: '',
+             scope: {inputA: '<'},
+             bindToController: false,
+             controllerAs: '$ctrl',
+             controller: class {
+               constructor($scope: angular.IScope) {
+                 $scope['$onChanges'] = scopeOnChangesA;
+                 (this as any).$onChanges = controllerOnChangesA;
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/test/aot/integration/upgrade_component_spec.ts]: what about testing `$onChanges` on the prototype too?

**Comment by petebacondarwin** [comment]
eighth commit 095313854bbb71a7e8954f63fbce89a5457e0c6b LGTM but could do with a test of `$onChanges` on the controller prototype too.

File: modules/@angular/upgrade/src/aot/upgrade_component.ts [hunk 11]
```diff
@@ -84,29 +84,23 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck {
...
+          `Upgraded directive '${name}' specifies 'bindToController' but no controller.`);
     }
 
+    this.bindingDestination = bindToController ? this.controllerInstance : this.$componentScope;
+
     this.setupOutputs();
   }
 
   ngOnInit() {
     // QUESTION: why not just use $compile instead of reproducing parts of it
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/aot/upgrade_component.ts]: We can remove this question now, I think.
The reasoning is that `$compile` needs to own the element that it is compiling, but in ngUpgrade, it is Angular 2 that owns the element on which we are compiling an Angular 1 upgraded component. So we must reproduce the compilation behaviour here.

File: modules/@angular/upgrade/src/aot/upgrade_component.ts [hunk 12]
```diff
@@ -84,29 +84,23 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck {
...
-      } else {
-        throw new Error(
-            `Upgraded directive '${name}' specifies 'bindToController' but no controller.`);
-      }
-    } else {
-      this.bindingDestination = this.$componentScope;
+    const bindToController = this.directive.bindToController;
+    if (controllerType) {
+      this.controllerInstance = this.buildController(
+          controllerType, this.$componentScope, this.$element, this.directive.controllerAs);
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/aot/upgrade_component.ts]: We should probably document somewhere that we are following the planned Angular 1.7 approach of never "pre-binding" the controller in ngUpgrade.

**Comment by petebacondarwin** [comment]
ninth commit 63bd3c2de3480d75db1346fde1cf07d110126cba LGTM

**Comment by petebacondarwin** [comment]
tenth commit 5c7c3382b6d63c6418ba8a7861ebca2072ec02ac LGTM but I agree that we should be responsible for destroying the `$componentScope`.

**Comment by petebacondarwin** [comment]
eleventh commit e02aaa313bf8fc2dac8b6769e59fb2601313a20e LGTM

**Comment by vicb** [comment]
what is the status of this PR ?

**Comment by petebacondarwin** [comment]
@vicb I couldn't find a label that indicated that there were a few small issues that I would like addressing before I would say that this is ready to merge.

File: modules/@angular/upgrade/src/angular_js.ts (hunk 1 above)
Comment (gkalpak) [inline modules/@angular/upgrade/src/angular_js.ts]: This change is not necessary, I just was more accurate.
AFAICT, it is incidental that `Ng1Token` is string and `DirectiveRequireProperty` accepts strings. Since required controllers are stored in `.data()`, they can only be retrieved by string (the key under which they are stored), while `Ng1Token` may be more than string in the future :wink:

File: modules/@angular/upgrade/test/aot/integration/upgrade_component_spec.ts (hunk 8 above)
Comment (gkalpak) [inline modules/@angular/upgrade/test/aot/integration/upgrade_component_spec.ts]: Yes, that's what ng1 does. If you don't require anything, you get your own controller (if you have one). You still need to require your own controller if you also need to require others.

File: modules/@angular/upgrade/src/angular_js.ts (hunk 1 above)
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/angular_js.ts]: Good point! +1

File: modules/@angular/upgrade/src/angular_js.ts [hunk 13]
```diff
@@ -12,6 +12,8 @@ export interface IAnnotatedFunction extends Function { $inject?: Ng1Token[]; }
 
 export type IInjectable = (Ng1Token | Function)[] | IAnnotatedFunction;
 
+export type SingleOrListOrMap<T> = T | T[] | {[key: string]: T};
```
Comment (gkalpak) [inline modules/@angular/upgrade/src/angular_js.ts]: Please suggest a better name :confused:

**Comment by gkalpak** [comment]
@petebacondarwin, PTAL.
I added a few commits - didn't touch the previous ones, so you only have to review the new ones.

File: modules/@angular/upgrade/test/aot/integration/upgrade_component_spec.ts [hunk 14]
```diff
@@ -2135,9 +2135,10 @@ export function main() {
              template: 'Called: {{ called }}',
              bindToController: true,
              controller: class {
-               constructor(private $scope: angular.IScope) { $scope['called'] = 'no'; }
-
-               $onInit() { this.$scope['called'] = 'yes'; }
+               constructor($scope: angular.IScope) {
+                 $scope['called'] = 'no';
+                 (this as any)['$onInit'] = () => $scope['called'] = 'yes';
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/test/aot/integration/upgrade_component_spec.ts]: Isn't this change moving it from the prototype to the instance?

File: modules/@angular/upgrade/test/aot/integration/upgrade_component_spec.ts [hunk 15]
```diff
@@ -2205,6 +2207,7 @@ export function main() {
                constructor($scope: angular.IScope) {
                  $scope['called'] = 'no';
                  $scope['$onInit'] = () => $scope['called'] = 'yes';
+                 Object.getPrototypeOf($scope)['$onInit'] = () => $scope['called'] = 'yes';
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/test/aot/integration/upgrade_component_spec.ts]: with this change and the change above the is `$onInit` on both the scope and the scope's prototype. So we can't tell which is being called. Was that the intention?

**Review by petebacondarwin — COMMENTED** [review]
52a390d73df8181893eeae4a307871092a23ccde looks good

**Review by petebacondarwin — COMMENTED** [review]
384f174253ab373859df6ac9e9f20dccd96532b7 LGTM

File: modules/@angular/upgrade/test/aot/integration/upgrade_component_spec.ts (hunk 15 above)
Comment (gkalpak) [inline modules/@angular/upgrade/test/aot/integration/upgrade_component_spec.ts]: We just want to verify that none is called, so we can have the same method.
I can change it if you think it is confusing.

**Comment by petebacondarwin** [comment]
LGTM - please rebase and squash. Then it is ready to merge

**Comment by vikerman** [comment]
Will merge after rebase

**Comment by gkalpak** [comment]
I rebased and squashed it down to 2 commits (one for lifecycle hooks, one for `require`).
Let me know if it is better to squash them into one.

**Comment by vikerman** [comment]
@vicb certified this new feature as not to be breaking. Merging...
